File: packages/python/dirsql/plan_entry_path.py
"""The loadable path one planned extension entry resolves to."""

from __future__ import annotations

import os

from .resolve_package import _resolve_package


def _plan_entry_path(entry):
    """Resolve a planned entry to a concrete file.

    A literal entry is already resolved. A package entry prefers a same-named
    file shadowing it next to the config, then the installed package.
    """
    if entry.path is not None:
        return entry.path
    if os.path.isfile(entry.shadow):
        return entry.shadow
    return _resolve_package(entry.package)

File: packages/ts/src/plan-entry-path.ts
// The loadable path one planned extension entry resolves to.

import { existsSync, statSync } from "node:fs";
import type { ExtensionPlanEntry } from "./core.js";
import { defaultResolver } from "./default-resolver.js";
import type { PackageResolver } from "./package-dir.js";
import { resolvePackage } from "./resolve-package.js";

/** Resolve a planned entry to a concrete file.
 *
 * A literal entry is already resolved. A package entry prefers a same-named
 * file shadowing it next to the config, then the installed package.
 */
export function planEntryPath(
  entry: ExtensionPlanEntry,
  resolver: PackageResolver = defaultResolver(),
): string {
  if (entry.path !== undefined) {
    return entry.path;
  }
  if (existsSync(entry.shadow) && statSync(entry.shadow).isFile()) {
    return entry.shadow;
  }
  return resolvePackage(entry.package, resolver);
}
